### DISCUSSION TITLE
onSuccess only on fetching

### DISCUSSION THREAD

**Discussion by dstoyanoff** [TanStack/query#4067] [General]
Hello,
I am trying to implement the following:

I have a query that gets data from the backend. When the data is resolved, I subscribe to a SignalR endpoint to get updates and modify the cache using `setQueryData`. The problem is that `setQueryData` would also call `onSuccess`, going into an infinite loop. Same with `onSettled`. Is there a place where I can attach only to the initial query (either network or resolved from the cache), but not to updates coming from manual cache updates?

**Comment by TkDodo** [comment]
> The problem is that setQueryData would also call onSuccess, going into an infinite loop.

This behaviour has been changed in v4:

https://tanstack.com/query/v4/docs/guides/migrating-to-react-query-4#onsuccess-is-no-longer-called-from-setquerydata

**Reply by TkDodo** [reply]
> Is there a place where I can attach only to the initial query (either network or resolved from the cache), but not to updates coming from manual cache updates?

So in v4, `onSuccess` is for network requests only. What you are describing does not really exist. Not sure why it's important to set up the subscription _after_ the data has been fetched for the first time? In my examples with websockets, I just setup the subscription separately: https://tkdodo.eu/blog/using-web-sockets-with-react-query

**Reply by dstoyanoff** [reply]
Thank you for the fast response. When the initial data is resolved, it would return the full objects from the list. When there is a signalr update, we only send deltas, which are just a few fields from the graph. If the data is not yet in the cache, I need to ignore deltas, which is doable, but not so optimal. But I might go with that approach for now, until I find a better way to do it. Thank you
